(PIE-281) Revive Splunk Acceptance Tests

gsparks opened the pull request:
This PR migrates much of the acceptance testing infrastructure from the
reporting module such that now a puppet master node is created and
installed on vmpooler with the splunk container running within it.

No new acceptance tests have been added. but the existing tests run once
again.

Additionally the splunk container is started and ready to accept HEC
requests for further acceptance testing. The Puppet Report Viewer
Splunkbase app is installed into the Splunk container so that we can now
send HEC requests to the container and verify that they were indexed
correctly.

A jenkins project has been created.
https://cinext-jenkinsmaster-pipeline-prod-1.delivery.puppetlabs.net/view/PIE%20Team/job/pipeline_puppetlabs-splunk_hec/

gsparks commented:
I'm not sure how we feel about having both Github actions and Jenkins running simultaneously, but I figured I'd leave it until we have consensus. Right now Actions handles the static checks / deploy process and jenkins handles the acceptance tests.

gsparks commented:
Also I figured I'd add additional acceptance tests later as this PR is large enough as it is.

ekinanp commented on Rakefile:
```diff
@@ -75,3 +75,123 @@ EOM
+    Rake::Task['litmus:install_module'].invoke(master.uri)
+  end
+
+  # splunk_hec.yaml must be deleted from the master if this is run more than once.
```
I'd go ahead and do this b/c I see no good reason to have the user do this.. Maybe `master.run_shell('rm -rf <splunk_hec.yaml>')`.

gsparks replied:
adding this to the `install_module` task, and removing the `configure_module` task.

ekinanp commented on spec/acceptance/class_spec.rb:
```diff
@@ -24,7 +24,7 @@ class { 'splunk_hec': url => "http://httpbin.org/post", token => '', record_even
 
     it 'Successfully sends data to an http endpoint' do
-      run_shell('cat /etc/puppetlabs/code/environments/production/modules/splunk_hec/examples/orchestrator_metrics.json | puppet splunk_hec --sourcetype puppet:summary --saved_report')
+      run_shell('cat /etc/puppetlabs/code/environments/production/modules/splunk_hec/examples/orchestrator_metrics.json | puppet splunk_hec --sourcetype puppet:summary')
```
Why this change?

gsparks replied:
The `--saved_report` option will submit the request immediately to the splunk instance configured in splunk_hec.yaml. It will not work currently as that file is not yet fully configured with the credentials for the splunk container we are running. I'm planning a PR right after this one to add this back and more acceptance tests that actually communicate with the Splunk container. I could also add it for just this test in this PR as well if we feel that changing this test feels janky. Happy to go either way.

ekinanp replied:
Ah ok, that's fine. Was just curious.

ekinanp commented on spec/support/acceptance/configure_splunk_hec.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Why is this necessary? Shouldn't testing the setup be included in the acceptance tests?

gsparks replied:
Turns out this isn't neccesary. Thanks!

ekinanp commented on spec/support/acceptance/splunk/docker-compose.yml:
```diff
@@ -0,0 +1,27 @@
+    hostname: splunk_instance
+    environment:
+      - SPLUNK_START_ARGS=--accept-license
+      - ASFDGASDFG=5
```
What is this variable for?

gsparks replied:
nothing, thanks! Will remove.

ekinanp commented on spec/support/acceptance/splunk/docker-compose.yml:
```diff
@@ -0,0 +1,27 @@
+      # Report Viewer from Splunkbase.
+      # We could alternatively download the packaged app from a location
+      # like github for testing.
+      - SPLUNK_APPS_URL=https://splunkbase.splunk.com/app/4413/release/3.0.1/download
```
This is a persistent instance right?

gsparks replied:
Yeah this is just a splunkbase account that is used to download splunk apps. It's separate from a login to a splunk instance.

gsparks replied:
I tried to create a team account, but team-pie@puppet.com already has an account. I'll update it as soon as I find the password haha.

ekinanp commented on spec/support/acceptance/splunk/default.yml:
```diff
@@ -0,0 +1,19 @@
+---
```
What's this file for?

gsparks replied:
This file is passed into the container on startup and is used to configure all the splunk settings you'd normally configure via the web interface.

[https://splunk.github.io/docker-splunk/ADVANCED.html](https://splunk.github.io/docker-splunk/ADVANCED.html)

RandomNoun7 replied:
Isn't this for configuring the splunk container?

ekinanp replied:
Ah ok. I'd maybe post a comment in the file along those lines. May be nice to link to a reference? E.g. something like https://splunk.github.io/docker-splunk/ADVANCED.html

gsparks replied:
will do!

RandomNoun7 commented on Gemfile:
```diff
@@ -25,15 +25,16 @@ group :development do
-end
-group :system_tests do
+  gem "rspec_junit_formatter",                                   require: false
+  gem "rspec-puppet", '= 2.7.10',                                require: false
```
I think this should be slightly more loosely pinned. Something like `>= 2.7.10 < 3` or whatever syntax like that works.

RandomNoun7 commented on spec/spec_helper_acceptance_local.rb:
```diff
@@ -0,0 +1,59 @@
+
+require 'serverspec'
+require 'puppet_litmus'
+require './spec/support/acceptance/helpers.rb'
```
I don't think you need the `'./'` part of the path. Paths under `spec` are automatically added to `$LOAD_PATH` when running rspec.

gsparks replied:
hmm, I'm getting a load error when I remove the `./`.

gsparks replied:
`support/acceptance/helpers.rb` seems to be working.

RandomNoun7 commented on spec/spec_helper_acceptance_local.rb:
```diff
@@ -0,0 +1,59 @@
+# TODO: This will cause some problems if we run the tests
+# in parallel. For example, what happens if two targets
+# try to modify site.pp at the same time?
+def set_sitepp_content(manifest)
```
Would be great to get this as a helper into Litmus after Enis' PR is merged. Just a thought.

gsparks replied:
Will plan to update once we get this into Litmus.

ekinanp replied:
We don't need the TODO anymore btw, that was there back before the multi-node Litmus crap.

RandomNoun7 commented on spec/support/acceptance/splunk/default.yml:
```diff
@@ -0,0 +1,19 @@
+    ssl: False
+    port: 8088
+    token: abcd1234
+  smartstore: null
```
nit: missing newline.

RandomNoun7 commented on spec/support/acceptance/splunk/docker-compose.yml:
```diff
@@ -0,0 +1,27 @@
+      # localhost:8000 will bring up the web interface
+      - "127.0.0.1:8000:8000"
+      # 8088 is the hec endpoint
+      - "127.0.0.1:8088:8088"
```
Changing these bindings to `0.0.0.0`  might make things easier later for things like troubleshooting the tests and prevent an annoying bug for someone later on.

RandomNoun7 commented on spec/support/acceptance/splunk_hec.yaml:
```diff
@@ -0,0 +1,5 @@
+"url" : "http://localhost:8088/services/collector/event"
+"token" : "abcd1234"
+"enable_reports" : "true"
+"manage_routes" : "true"
```
nit: Newline

RandomNoun7 commented:
You can try just support/acceptance/blah, and if that’s still broken then
disregard completely because it’s not worth putting a lot of effort into
troubleshooting.

On Tue, Jul 28, 2020 at 15:23 Greg Sparks <notifications@github.com> wrote:

> *@gsparks* commented on this pull request.
> ------------------------------
>
> In spec/spec_helper_acceptance_local.rb
> <https://github.com/puppetlabs/puppetlabs-splunk_hec/pull/50#discussion_r461927390>
> :
>
> > @@ -0,0 +1,59 @@
> +# rubocop:disable Style/AccessorMethodName
> +
> +require 'serverspec'
> +require 'puppet_litmus'
> +require './spec/support/acceptance/helpers.rb'
>
> hmm, I'm getting a load error when I remove the ./.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-splunk_hec/pull/50#discussion_r461927390>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADEJN4OGFVNFSNU6DREMGZDR55FW7ANCNFSM4OY4CGSA>
> .
>

gsparks commented:
Fixed an issue where the `--saved_report `argument for the report processor caused the report to fail because the report data did not include a sourcetype which is used to specify different HEC tokens for different puppet report types eg.(summary, facts). Sourcetype should not always be required for this kind of testing purpose.

It looks like this test probably never ran correctly given the sample json report data that is in the repo.

Added a readme snippet about running the tests.

ekinanp commented on lib/puppet/util/splunk_hec.rb:
```diff
@@ -45,10 +45,17 @@ def create_http(source_type)
+      http = create_http(source_type)
+      token = settings[token_name] || settings['token'] || raise(Puppet::Error, 'Must provide token parameter to splunk class')
+    else
+      # For testing purposes there isn't always a sourcetype.
```
Seems weird that this is only here to get the tests to pass. Is it an important precondition of this module that the `sourcetype` must exist? If so, may be worth finding a way to get that test to pass _without_ having this fix. That way can be done in a follow-up PR though.

gsparks replied:
Splunk should be able to receive data without a specific sourcetype although the module isn't currently set up to do this automatically. It's nice to be able to do this for testing purposes, however an alternative approach would be to specify a sourcetype which does not exist, which creates the same behavior. That would be a way to keep this test, without modifying the production code.

ekinanp replied:
Yeah, let's do the "nonexistent sourcetype" option then if it's easy enough.

gsparks replied:
Cool, updated!
